src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,845 @@
+    """
+    Given a list of hidden states, batch them together using torch.stack.
+    We assume that the hidden states are of shape (batch_size, hidden_state_size, num_keypoints).
+    The list can be of two forms:
+    - List of hidden states: [hidden_state_0, hidden_state_1, ...]
+    - List of tuples of hidden states: [(hidden_state_0_0, hidden_state_1_0, ...), (hidden_state_0_1, hidden_state_1_1, ...), ...]
+    If the list is of the first form, we stack the attention probabilities along the batch dimension:
+    [hidden_state_0, hidden_state_1, ...] -> torch.stack([hidden_state_0, hidden_state_1, ...], dim=0)
+    If the list is of the second form, we stack the attention probabilities along the batch dimension for each tuple:
+    [(hidden_state_0_0, hidden_state_1_0, ...), (hidden_state_0_1, hidden_state_1_1, ...), ...] ->
+    [torch.stack([hidden_state_0_0, hidden_state_1_0, ...], dim=0), torch.stack([hidden_state_0_1, hidden_state_1_1, ...], dim=0), ...]
+    """
+    if len(hidden_states) > 0:
+        if isinstance(hidden_states[0], torch.Tensor):
+            return stack_hidden_states_list(hidden_states)
+        elif isinstance(hidden_states[0], tuple):
+            list_of_tuples = [tuple([element[i] for element in hidden_states]) for i in range(len(hidden_states[0]))]
+            return [stack_hidden_states_list(element) for element in list_of_tuples]
+        elif hidden_states[0] is None:
+            return None
+    else:
+        return []
+
+
+def attention(query: torch.Tensor, key: torch.Tensor, value: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Applied attention mechanism to query, key and value."""
+    dim = query.shape[1]
+    scores = torch.einsum("bdhn,bdhm->bhnm", query, key) / dim**0.5
+    prob = torch.nn.functional.softmax(scores, dim=-1)
+    output = torch.einsum("bhnm,bdhm->bdhn", prob, value)
+    return output, prob
+
+
+def normalize_keypoints(keypoints: torch.Tensor, height: int, width: int):
+    """Normalize keypoints locations based on image image_shape"""
+    one = keypoints.new_tensor(1)
+    size = torch.stack([one * width, one * height])[None]
+    center = size / 2
+    scaling = size.max(1, keepdim=True).values * 0.7
+    return (keypoints - center[:, None, :]) / scaling[:, None, :]
+
+
+def log_sinkhorn_iterations(
+    Z: torch.Tensor, log_mu: torch.Tensor, log_nu: torch.Tensor, iterations: int
+) -> torch.Tensor:
+    """Perform Sinkhorn Normalization in Log-space for stability"""
+    u, v = torch.zeros_like(log_mu), torch.zeros_like(log_nu)
+    for _ in range(iterations):
+        u = log_mu - torch.logsumexp(Z + v.unsqueeze(1), dim=2)
+        v = log_nu - torch.logsumexp(Z + u.unsqueeze(2), dim=1)
+    return Z + u.unsqueeze(2) + v.unsqueeze(1)
+
+
+def log_optimal_transport(scores: torch.Tensor, alpha: torch.Tensor, iterations: int) -> torch.Tensor:
+    """Perform Differentiable Optimal Transport in Log-space for stability"""
+    b, m, n = scores.shape
```

Comment:
We should use more descriptive names that single letter e.g. `batch_size` instead of `b`. Same for all other functions

Reply:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/29886/commits/601da1cf0756cef7339d4dd33faeefbfcd073ac9)